### PR TITLE
trigger by tag group

### PR DESCRIPTION
**Pull request by mikekg**
# trigger a set of plugins defined by a tag

## Overview
This branch implements component grouping via tags and significantly improves test coverage amd documentation. 

### 1. Test Coverage Improvements
- Added comprehensive test coverage for plugin validation
- Added tests for nil Steps in plugins
- Regularized test case naming conventions
- Expanded test coverage for component list parsing
- Added tests for various edge cases and error conditions

### 2. Component List Parsing and Validation
- Improved component list parsing logic
- Enhanced validation of component list entries
- Fixed handling of run modes and parameters
- Added better error messages for invalid formats
- Improved parameter inheritance logic

### 3. Trigger by Tag Implementation
- Added initial implementation of component triggering by tags
- Added REST API endpoint for triggering components
- Implemented component grouping functionality
- Added basic error handling and response formatting

### 4. Documentation
- Updated all related function names and variables
- Added comprehensive documentation in PLUGIN.md
- Added examples for Slurm integration
- Updated API documentation and examples

### 1. `pkg/custom-plugins/spec_test.go`
- Major expansion of test coverage
- New test cases for all functionality
- Added tests for tag-based grouping
- Enhanced validation tests

### 2. `docs/PLUGIN.md` 
- Added new documentation sections
- Added examples and use cases
- Updated API documentation
- Added Slurm integration guide

### 3. `pkg/custom-plugins/spec.go`
- Enhanced component list parsing
- Improved validation logic
- Added tag handling
- Better error handling

### 4. `pkg/server/handlers_components.go`
- Added trigger by tag functionality
- Enhanced REST API endpoints
- Improved error handling
- Better response formatting

### 5. `pkg/custom-plugins/types.go`
- Updated field names
- Modified JSON tags
- Updated documentation

## Key Improvements
1. **Test Coverage and Reliability**
   - Comprehensive test suite
   - Edge case handling
   - Validation improvements

2. **Component List Parsing**
   - Enhanced parsing logic
   - Better validation
   - Improved error handling

3. **Component Grouping**
   - Tags-based grouping and merge with tags support from 778
   - REST API support
   - Trigger functionality

4. **Documentation**
   - Comprehensive guides
   - API documentation
   - Integration examples

5. **API Usability**
   - Enhanced endpoints
   - Better error handling
   - Improved response formatting

## Integration Examples
Added comprehensive examples for:
- Slurm prologue/epilogue integration
- Continuous monitoring
- Component grouping
- API usage
- Error handling

### CODE REVIEW

File: pkg/server/handlers_components.go [hunk 1]
```diff
@@ -607,3 +609,61 @@ func (g *globalHandler) getMetrics(c *gin.Context) {
...
+
+	for _, comp := range components {
+		// Check if component has the specified tag
+		// For now, we'll do a linear scan through all components
+		// This could be optimized with a tag-based index structure
+		if spec, ok := comp.(pkgcustomplugins.CustomPluginRegisteree); ok {
+			hasTag := false
+			for _, tag := range spec.Spec().Tags {
+				if tag == tagName {
+					hasTag = true
```
[the remark this answers is not in the document]
Comment (gyuho) [reply pkg/server/handlers_components.go:639]: @mikekg Now https://github.com/leptonai/gpud/pull/778 supports component.Tags method. So, we can do:

```go
tags := comp.Tags()
if match(tags, tagName) { ... }
```

(still needs rebase off that branch + spec.Tags needs to be used in custom plugin package)

[the remark this answers is not in the document]
Comment (gyuho) [reply pkg/server/handlers_components.go:639]: @mikekg Now that https://github.com/leptonai/gpud/pull/778 has been merged, we can derive the component tags from the interface.

**Review by gyuho — APPROVED** [review]
lgtm!!
